File: sdk/approute-sdk/javascript/tests/resources/funds.test.ts
import { describe, it, expect } from "vitest";
import { FundsResource } from "../../src/resources/funds.js";
import { MockTransport } from "../support/index.js";
import {
  makeFundingMethodsResponse,
  makeFundingInvoice,
  makeFundingInvoiceList,
  makeFundingInvoiceTimeLeft,
  makeTonDepositState,
  makeBybitState,
} from "../factories/index.js";
import type {
  FundingMethodsResponse,
  FundingInvoice,
  FundingInvoiceList,
  FundingInvoiceTimeLeft,
  TonDepositState,
  BybitState,
} from "../../src/models/index.js";

describe("FundsResource", () => {
  describe("methods()", () => {
    it("should return funding methods on success", async () => {
      const fixture = makeFundingMethodsResponse();
      const mock = new MockTransport(fixture);
      const resource = new FundsResource(mock.transport);

      const result: FundingMethodsResponse = await resource.methods();

      expect(result.items).toHaveLength(1);
      expect(result.items[0].code).toBe("USDT_TRC20");
      expect(result.items[0].name).toBe("USDT (TRC-20)");
      expect(result.items[0].minAmount).toBe(10.0);
      expect(result.items[0].commission).toBe(0.0);
      expect(result.items[0].ttlMinutes).toBe(60);
      expect(result.items[0].confirmationsRequired).toBe(20);
    });

    it("should call GET /funds/methods", async () => {
      const mock = new MockTransport(makeFundingMethodsResponse());
      const resource = new FundsResource(mock.transport);

      await resource.methods();

      expect(mock.callCount).toBe(1);
      expect(mock.lastCall.method).toBe("GET");
      expect(mock.lastCall.path).toBe("/funds/methods");
    });
  });

  describe("createInvoice()", () => {
    it("should return created invoice on success", async () => {
      const fixture = makeFundingInvoice();
      const mock = new MockTransport(fixture);
      const resource = new FundsResource(mock.transport);

      const result: FundingInvoice = await resource.createInvoice({
        methodCode: "USDT_TRC20",
        amount: 100,
      });

      expect(result.id).toBe("inv-001");
      expect(result.methodCode).toBe("USDT_TRC20");
      expect(result.amountExpected).toBe(100.0);
      expect(result.status).toBe("pending");
      expect(result.address).toBe("TXyz123abc");
      expect(result.direction).toBe("incoming");
    });

    it("should call POST /funds/invoices with correct body", async () => {
      const mock = new MockTransport(makeFundingInvoice());
      const resource = new FundsResource(mock.transport);

      await resource.createInvoice({
        methodCode: "USDT_TRC20",
        amount: 100,
      });

      expect(mock.callCount).toBe(1);
      expect(mock.lastCall.method).toBe("POST");
      expect(mock.lastCall.path).toBe("/funds/invoices");
      expect(mock.lastCall.options?.body).toEqual({
        methodCode: "USDT_TRC20",
        amount: 100,
      });
    });
  });

  describe("listInvoices()", () => {
    it("should return invoice list on success", async () => {
      const fixture = makeFundingInvoiceList();
      const mock = new MockTransport(fixture);
      const resource = new FundsResource(mock.transport);

      const result: FundingInvoiceList = await resource.listInvoices();

      expect(result.items).toHaveLength(1);
      expect(result.total).toBe(1);
      expect(result.items[0].id).toBe("inv-001");
    });

    it("should call GET /funds/invoices with default pagination", async () => {
      const mock = new MockTransport(makeFundingInvoiceList());
      const resource = new FundsResource(mock.transport);

      await resource.listInvoices();

      expect(mock.lastCall.method).toBe("GET");
      expect(mock.lastCall.path).toBe("/funds/invoices");
      expect(mock.lastCall.options?.params).toEqual({
        limit: 50,
        offset: 0,
      });
    });

    it("should pass all filter options", async () => {
      const mock = new MockTransport(makeFundingInvoiceList());
      const resource = new FundsResource(mock.transport);

      await resource.listInvoices({
        status: ["pending", "confirming"],
        methodCode: ["USDT_TRC20"],
        search: "inv-001",
        invoiceId: "inv-001",
        createdFrom: "2026-01-01",
        createdTo: "2026-03-01",
        withTx: true,
        limit: 10,
        offset: 0,
      });

      expect(mock.lastCall.options?.params).toEqual({
        status: ["pending", "confirming"],
        methodCode: ["USDT_TRC20"],
        search: "inv-001",
        invoiceId: "inv-001",
        createdFrom: "2026-01-01",
        createdTo: "2026-03-01",
        withTx: true,
        limit: 10,
        offset: 0,
      });
    });
  });

  describe("getInvoice()", () => {
    it("should return a single invoice on success", async () => {
      const fixture = makeFundingInvoice();
      const mock = new MockTransport(fixture);
      const resource = new FundsResource(mock.transport);

      const result: FundingInvoice = await resource.getInvoice("inv-001");

      expect(result.id).toBe("inv-001");
      expect(result.status).toBe("pending");
    });

    it("should call GET /funds/invoices/:id", async () => {
      const mock = new MockTransport(makeFundingInvoice());
      const resource = new FundsResource(mock.transport);

      await resource.getInvoice("inv-001");

      expect(mock.lastCall.method).toBe("GET");
      expect(mock.lastCall.path).toBe("/funds/invoices/inv-001");
    });
  });

  describe("checkInvoice()", () => {
    it("should call POST /funds/invoices/:id/check", async () => {
      const mock = new MockTransport(makeFundingInvoice());
      const resource = new FundsResource(mock.transport);

      await resource.checkInvoice("inv-001");

      expect(mock.lastCall.method).toBe("POST");
      expect(mock.lastCall.path).toBe("/funds/invoices/inv-001/check");
    });
  });

  describe("invoiceTimeLeft()", () => {
    it("should return time left on success", async () => {
      const fixture = makeFundingInvoiceTimeLeft();
      const mock = new MockTransport(fixture);
      const resource = new FundsResource(mock.transport);

      const result: FundingInvoiceTimeLeft =
        await resource.invoiceTimeLeft("inv-001");

      expect(result.invoiceId).toBe("inv-001");
      expect(result.secondsLeft).toBe(3200);
      expect(result.expired).toBe(false);
    });

    it("should call GET /funds/invoices/:id/time-left", async () => {
      const mock = new MockTransport(makeFundingInvoiceTimeLeft());
      const resource = new FundsResource(mock.transport);

      await resource.invoiceTimeLeft("inv-001");

      expect(mock.lastCall.method).toBe("GET");
      expect(mock.lastCall.path).toBe("/funds/invoices/inv-001/time-left");
    });
  });

  describe("tonDeposit()", () => {
    it("should return TON deposit state on success", async () => {
      const fixture = makeTonDepositState();
      const mock = new MockTransport(fixture);
      const resource = new FundsResource(mock.transport);

      const result: TonDepositState = await resource.tonDeposit();

      expect(result.address).toBe(
        "EQDtFpEwcFAEcRe5mLVh2N6C0x-_hJEM7W61_JLnSF78p7Av",
      );
      expect(result.memoTag).toBe("123456");
    });

    it("should call GET /funds/ton/deposit", async () => {
      const mock = new MockTransport(makeTonDepositState());
      const resource = new FundsResource(mock.transport);

      await resource.tonDeposit();

      expect(mock.lastCall.method).toBe("GET");
      expect(mock.lastCall.path).toBe("/funds/ton/deposit");
    });
  });

  describe("bybitState()", () => {
    it("should return Bybit state on success", async () => {
      const fixture = makeBybitState();
      const mock = new MockTransport(fixture);
      const resource = new FundsResource(mock.transport);

      const result: BybitState = await resource.bybitState();

      expect(result.recipientUid).toBe("bybit-uid-001");
      expect(result.linked).toBe(true);
      expect(result.yourUid).toBe("bybit-uid-002");
    });

    it("should call GET /funds/bybit/state", async () => {
      const mock = new MockTransport(makeBybitState());
      const resource = new FundsResource(mock.transport);

      await resource.bybitState();

      expect(mock.lastCall.method).toBe("GET");
      expect(mock.lastCall.path).toBe("/funds/bybit/state");
    });
  });

  describe("bybitAttach()", () => {
    it("should call POST /funds/bybit/attach with uid in body", async () => {
      const mock = new MockTransport(makeBybitState());
      const resource = new FundsResource(mock.transport);

      await resource.bybitAttach("bybit-uid-002");

      expect(mock.lastCall.method).toBe("POST");
      expect(mock.lastCall.path).toBe("/funds/bybit/attach");
      expect(mock.lastCall.options?.body).toEqual({ uid: "bybit-uid-002" });
    });
  });

  describe("bybitUnlink()", () => {
    it("should call POST /funds/bybit/unlink", async () => {
      const mock = new MockTransport(makeBybitState());
      const resource = new FundsResource(mock.transport);

      await resource.bybitUnlink();

      expect(mock.lastCall.method).toBe("POST");
      expect(mock.lastCall.path).toBe("/funds/bybit/unlink");
    });
  });
});

File: sdk/approute-sdk/golang/funds_test.go
package approute_test

import (
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"

	approute "github.com/approute/public-api-sdk-go"
	"github.com/approute/public-api-sdk-go/apierror"
	"github.com/approute/public-api-sdk-go/enum"
	"github.com/approute/public-api-sdk-go/model"
	"github.com/approute/public-api-sdk-go/resource"
)

func TestFundsResource_Methods(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "GET")
		assertEqual(t, "Path", r.URL.Path, "/funds/methods")
		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newFundingMethodsResponse(
			newFundingMethod(),
			newFundingMethod(func(m *model.FundingMethod) {
				m.Code = enum.FundingUSDT_TON
				m.Name = "USDT (TON)"
				m.MinAmount = 5.00
				m.Address = "UQAbc...xyz"
				m.TTLMinutes = 30
				m.ConfirmationsRequired = 1
			}),
		)))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	result, err := client.Funds.Methods(t.Context())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(result.Items) != 2 {
		t.Fatalf("expected 2 methods, got %d", len(result.Items))
	}
	assertEqual(t, "Items[0].Code", string(result.Items[0].Code), "USDT_TRC20")
	assertEqual(t, "Items[0].Name", result.Items[0].Name, "USDT (TRC-20)")
	assertEqual(t, "Items[0].MinAmount", result.Items[0].MinAmount, 10.00)
	assertEqual(t, "Items[1].Code", string(result.Items[1].Code), "USDT_TON")
}

func TestFundsResource_CreateInvoice(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "POST")
		assertEqual(t, "Path", r.URL.Path, "/funds/invoices")

		var body map[string]any
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("failed to decode request body: %v", err)
		}
		if body["methodCode"] != "USDT_TRC20" {
			t.Errorf("expected methodCode=USDT_TRC20, got %v", body["methodCode"])
		}

		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newFundingInvoice()))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	invoice, err := client.Funds.CreateInvoice(t.Context(), &model.FundingInvoiceCreateRequest{
		MethodCode: enum.FundingUSDT_TRC20,
		Amount:     50.00,
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	assertEqual(t, "ID", invoice.ID, "inv-001")
	assertEqual(t, "MethodCode", string(invoice.MethodCode), "USDT_TRC20")
	assertEqual(t, "AmountExpected", invoice.AmountExpected, 50.00)
	assertEqual(t, "Status", string(invoice.Status), "pending")
}

func TestFundsResource_ListInvoices(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "GET")
		assertEqual(t, "Path", r.URL.Path, "/funds/invoices")

		q := r.URL.Query()
		if q.Get("limit") == "" {
			t.Error("expected limit parameter")
		}

		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newFundingInvoiceList(
			newFundingInvoice(func(inv *model.FundingInvoice) {
				inv.ConfirmationsRequired = nil
				inv.Confirmations = nil
			}),
		)))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	result, err := client.Funds.ListInvoices(t.Context(), nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	assertEqual(t, "Total", result.Total, 1)
	if len(result.Items) != 1 {
		t.Fatalf("expected 1 invoice, got %d", len(result.Items))
	}
	assertEqual(t, "Items[0].ID", result.Items[0].ID, "inv-001")
}

func TestFundsResource_ListInvoices_WithOptions(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()
		assertEqual(t, "status", q.Get("status"), "pending,success")
		assertEqual(t, "methodCode", q.Get("methodCode"), "USDT_TRC20")
		assertEqual(t, "limit", q.Get("limit"), "10")

		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newFundingInvoiceList(
			newFundingInvoice(func(inv *model.FundingInvoice) {
				inv.ConfirmationsRequired = nil
				inv.Confirmations = nil
			}),
		)))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	_, err := client.Funds.ListInvoices(t.Context(), &resource.InvoiceListOptions{
		Status:     []enum.FundingStatus{enum.FundingPending, enum.FundingSuccess},
		MethodCode: []enum.FundingMethodCode{enum.FundingUSDT_TRC20},
		Limit:      10,
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestFundsResource_GetInvoice(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "GET")
		assertEqual(t, "Path", r.URL.Path, "/funds/invoices/inv-001")
		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newFundingInvoice(func(inv *model.FundingInvoice) {
			inv.Credited = 50.00
			inv.TxHash = strPtr("0xabc123")
			inv.Status = enum.FundingSuccess
			inv.ConfirmationsRequired = intPtr(20)
			inv.Confirmations = intPtr(20)
		})))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	invoice, err := client.Funds.GetInvoice(t.Context(), "inv-001")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	assertEqual(t, "ID", invoice.ID, "inv-001")
	assertEqual(t, "Status", string(invoice.Status), "success")
	assertEqual(t, "Credited", invoice.Credited, 50.00)
	if invoice.TxHash == nil {
		t.Fatal("TxHash should not be nil")
	}
	assertEqual(t, "TxHash", *invoice.TxHash, "0xabc123")
}

func TestFundsResource_CheckInvoice(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "POST")
		assertEqual(t, "Path", r.URL.Path, "/funds/invoices/inv-001/check")
		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newFundingInvoice(func(inv *model.FundingInvoice) {
			inv.Status = enum.FundingConfirming
			inv.ConfirmationsRequired = intPtr(20)
			inv.Confirmations = intPtr(5)
		})))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	invoice, err := client.Funds.CheckInvoice(t.Context(), "inv-001")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	assertEqual(t, "Status", string(invoice.Status), "confirming")
	if invoice.Confirmations == nil {
		t.Fatal("Confirmations should not be nil")
	}
	assertEqual(t, "Confirmations", *invoice.Confirmations, 5)
}

func TestFundsResource_InvoiceTimeLeft(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "GET")
		assertEqual(t, "Path", r.URL.Path, "/funds/invoices/inv-001/time-left")
		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newFundingInvoiceTimeLeft()))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	result, err := client.Funds.InvoiceTimeLeft(t.Context(), "inv-001")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	assertEqual(t, "InvoiceID", result.InvoiceID, "inv-001")
	assertEqual(t, "SecondsLeft", result.SecondsLeft, 1800)
	assertEqual(t, "Expired", result.Expired, false)
}

func TestFundsResource_TonDeposit(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "GET")
		assertEqual(t, "Path", r.URL.Path, "/funds/ton/deposit")
		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newTonDepositState()))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	result, err := client.Funds.TonDeposit(t.Context())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	assertEqual(t, "Address", result.Address, "UQAbc...xyz")
	assertEqual(t, "MemoTag", result.MemoTag, "12345")
}

func TestFundsResource_BybitState(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "GET")
		assertEqual(t, "Path", r.URL.Path, "/funds/bybit/state")
		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newBybitState()))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	result, err := client.Funds.BybitState(t.Context())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	assertEqual(t, "RecipientUID", result.RecipientUID, "bybit-uid-001")
	assertEqual(t, "Linked", result.Linked, true)
	if result.YourUID == nil {
		t.Fatal("YourUID should not be nil")
	}
	assertEqual(t, "YourUID", *result.YourUID, "my-uid-001")
}

func TestFundsResource_BybitAttach(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "POST")
		assertEqual(t, "Path", r.URL.Path, "/funds/bybit/attach")

		var body map[string]any
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("failed to decode request body: %v", err)
		}
		if body["uid"] != "new-uid-002" {
			t.Errorf("expected uid=new-uid-002, got %v", body["uid"])
		}

		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newBybitState(func(s *model.BybitState) {
			s.YourUID = strPtr("new-uid-002")
		})))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	result, err := client.Funds.BybitAttach(t.Context(), "new-uid-002")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	assertEqual(t, "Linked", result.Linked, true)
	if result.YourUID == nil {
		t.Fatal("YourUID should not be nil")
	}
	assertEqual(t, "YourUID", *result.YourUID, "new-uid-002")
}

func TestFundsResource_BybitUnlink(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "POST")
		assertEqual(t, "Path", r.URL.Path, "/funds/bybit/unlink")
		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newBybitState(func(s *model.BybitState) {
			s.Linked = false
			s.YourUID = nil
		})))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	result, err := client.Funds.BybitUnlink(t.Context())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	assertEqual(t, "Linked", result.Linked, false)
	if result.YourUID != nil {
		t.Error("YourUID should be nil after unlink")
	}
}

func TestFundsResource_GetInvoice_NotFound(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(404)
		w.Header().Set("Content-Type", "application/json")
		w.Write(errorEnvelope(t, "NOT_FOUND", "Invoice not found", "t-fund-err-1"))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	_, err := client.Funds.GetInvoice(t.Context(), "nonexistent")
	if err == nil {
		t.Fatal("expected error, got nil")
	}

	var nf *apierror.NotFoundError
	if !errors.As(err, &nf) {
		t.Fatalf("expected *NotFoundError, got %T: %v", err, err)
	}
	assertEqual(t, "Code", nf.Code, "NOT_FOUND")
}
